scrapy/dispatch/dispatcher.py

```diff
@@ -0,0 +1,390 @@
+        self.use_caching = use_caching
+        # For convenience we create empty caches even if they are not used.
+        # A note about caching: if use_caching is defined, then for each
+        # distinct sender we cache the receivers that sender has in
+        # 'sender_receivers_cache'. The cache is cleaned when .connect() or
+        # .disconnect() is called and populated on send().
+        self.sender_receivers_cache = weakref.WeakKeyDictionary(
+        ) if use_caching else {}
+        self.receiver_accepts_kwargs = {}
+        self._dead_receivers = False
+
+    def connect(self, receiver, sender=None, weak=True, dispatch_uid=None):
+        """
+        Connect receiver to sender for signal.
+
+        :param receiver: A function or an instance method which is
+                         to receive signals. If a receiver is connected
+                         with a dispatch_uid argument, it will not be added
+                         if another receiver was already connected that
+                         dispatch_uid.
+        :type receiver: function or instance, must be hashable object with
+
+        :param sender: The sender to which the receiver should respond.
+                       Must either be a Python object, or None to receive
+                       events from any sender.
+        :type sender: object, None
+
+        :param weak: Whether to use weak references to the receiver. By
+                     default, the module will attempt to use weak
+                     references to the receiver objects. If this parameter
+                     is false, then strong references will be used.
+        :type weak: Boolean
+
+        :param dispatch_uid: An identifier used to uniquely identify a
+                             particular instance of a receiver.
+        :type dispatch_uid: string, though it may be anything
+                            hashable.
+        """
+        assert callable(receiver), "Signal receivers must be callable."
+        # Check for **kwargs
+        accepts_kwargs = True
+        if not func_accepts_kwargs(receiver):
+            warnings.warn("The use of handlers that don't accept "
+                          "**kwargs has been deprecated, plese refer "
+                          "to the Signals API documentation.",
+                          ScrapyDeprecationWarning, stacklevel=3)
+            accepts_kwargs = False
+        if dispatch_uid:
+            lookup_key = (dispatch_uid, _make_id(sender))
+        else:
+            lookup_key = (_make_id(receiver), _make_id(sender))
+        accepts_kwargs_lookup = _make_id(receiver)
+        if weak:
+            ref = weakref.ref
+            receiver_object = receiver
+            # Check for bound methods
+            if hasattr(receiver, '__self__') and hasattr(receiver, '__func__'):
+                ref = WeakMethod
+                receiver_object = receiver.__self__
+            if six.PY3 and sys.version_info >= (3, 4):
```

Comment:
why is six.PY3 check needed? It may return False in Python 4.x, and it looks redundant because of sys.version_info check.